src/main/java/ca/uhn/fhir/jpa/starter/MessagePlainProvider.java

```diff
@@ -0,0 +1,167 @@
+    }
+
+    // check that there is a message header
+    Resource firstResource = bundleR4.getEntryFirstRep().getResource();
+    if ( firstResource.getResourceType() != ResourceType.MessageHeader) {
+      OperationOutcome oo = new OperationOutcome();
+      oo.addIssue()
+          .setSeverity(OperationOutcome.IssueSeverity.ERROR)
+          .setDiagnostics("First resource in message bundle must be of type 'MessageHeader'");
+      return oo;
+    }
+
+    MessageHeader msgHead = (MessageHeader) firstResource;
+
+    // check messageHeader is well formed
+    if (!msgHead.hasSource()) {
+      OperationOutcome oo = new OperationOutcome();
+      oo.addIssue()
+          .setSeverity(OperationOutcome.IssueSeverity.ERROR)
+          .setDiagnostics("'MessageHeader' is missing source");
+      return oo;
+    } else if (!msgHead.getSource().hasEndpoint()){
+      OperationOutcome oo = new OperationOutcome();
+      oo.addIssue()
+          .setSeverity(OperationOutcome.IssueSeverity.ERROR)
+          .setDiagnostics("'MessageHeader.source' is missing an endpoint");
+      return oo;
+    }
+
+    IFhirResourceDao bundleDao = daoRegistry.getResourceDao(ResourceType.Bundle.name());
+    bundleDao.create(bundleR4);
+
+
+    if (async != null && async.equals("true")) {
+      // asynchronous
+      String asyncResponseUrl;
+      if (responseUrl != null) {
+        asyncResponseUrl = responseUrl;
+      } else {
+        asyncResponseUrl = msgHead.getSource().getEndpoint();
+      }
+      // do any async operations in the new thread
+      Thread newThread = new Thread(() -> {
+        ResponseType responseType = doEvent(msgHead);
+        Bundle bundle = buildResponseMessage(msgHead, responseType);
+        IParser parser = fhirContext.newJsonParser();
+        String bundleString = parser.encodeResourceToString(bundle);
+        makePost(asyncResponseUrl, bundleString);
+      });
+      newThread.start();
+      OperationOutcome oo = new OperationOutcome();
+      oo.addIssue()
+          .setSeverity(OperationOutcome.IssueSeverity.INFORMATION)
+          .setDiagnostics("Asynchronous message successfully received");
+      return oo;
+
+    } else {
+      // synchronous
+      ResponseType responseType = doEvent(msgHead);
+      return buildResponseMessage(msgHead, responseType);
+    }
+  }
+
+  private void makePost(String asyncResponseUrl, String content) {
+    HttpClient httpClient = HttpClients.createDefault();
+    HttpPost httpPost = new HttpPost(asyncResponseUrl);
+    try {
+      StringEntity entity = new StringEntity(content);
+      httpPost.setEntity(entity);
+      httpPost.setHeader("Content-type", "application/json");
+      httpClient.execute(httpPost);
+    } catch (UnsupportedEncodingException e) {
+      e.printStackTrace();
+    } catch (ClientProtocolException e) {
+      e.printStackTrace();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+  private ResponseType doEvent(MessageHeader msgHead) {
```

Comment:
This seems fine for now, but I'm wondering if there's a good way to make this extensible. We're potentially going to want  different actors to support different actions. Not worth implementing anything else now, but worth thinking about